File: 42.trapping-rain-water.go
/*
 * @lc app=leetcode id=42 lang=golang
 *
 * [42] Trapping Rain Water
 */

// @lc code=start
func trap(height []int) int {
	res, left, right, leftmax, rightmax := 0, 0, len(height)-1, 0, 0

    if len(height) <= 2{
        return 0
    }

	for left < right {
		if height[left] <= height[right] {
			if height[left] > leftmax {
				leftmax = height[left]
			} else {
				res += leftmax - height[left]
			}

            left++
		} else {
			if height[right] > rightmax {
				rightmax = height[right]
			} else {
				res += rightmax - height[right]
			}

            right--
		}
	}

    return res
}
// @lc code=end

File: 42.trapping-rain-water.cpp
/*
 * @lc app=leetcode id=42 lang=cpp
 *
 * [42] Trapping Rain Water
 */

// @lc code=start
class Solution {
 public:
  int trap(vector<int>& height) {
    stack<int> st;
    int sum = 0;
    for (int i = 0; i < height.size(); ++i) {
      while (!st.empty() && height[i] > height[st.top()]) {
        int mid = st.top();
        st.pop();

        if (!st.empty()) {
          int h = min(height[st.top()], height[i]) - height[mid];
          int w = i - st.top() - 1;
          sum += h * w;
        }
      }
      st.push(i);
    }

    return sum;
  }
};
// @lc code=end

